File: core/src/com/sheepsgohome/leaderboard/LeaderBoard.java
package com.sheepsgohome.leaderboard;

import com.badlogic.gdx.Gdx;
import com.badlogic.gdx.Net;
import com.badlogic.gdx.net.HttpParametersUtils;
import com.badlogic.gdx.utils.Json;
import com.sheepsgohome.GameData;

import java.math.BigInteger;
import java.security.MessageDigest;
import java.security.NoSuchAlgorithmException;
import java.util.HashMap;
import java.util.Map;

/**
 * Created by LittleLight on 22.3.2015.
 */
public class LeaderBoard {
    public static final String RESPONSE_DB_CONNECTION_ERROR = "DB_CONNECTION_ERROR";
    public static final String RESPONSE_INVALID_DATA = "INVALID_DATA";
    public static final String RESPONSE_NICK_ALREADY_IN_USE = "NICK_ALREADY_IN_USE";
    public static final String RESPONSE_SUCCESS = "SUCCESS";
    public static final String RESPONSE_FAILURE = "FAILURE";
    public static final String RESPONSE_UNREGISTERED_USER = "UNREGISTERED_USER";

    private boolean terminated;


    public void fetchLeaderboard(String id, final LeaderBoardCallback callback) {
        terminated = false;

        Map parameters = new HashMap();
        parameters.put("id", id);

        Net.HttpRequest httpPost = new Net.HttpRequest(Net.HttpMethods.POST);
        httpPost.setUrl("http://sheepsgohome-com.svethostingu-tmp.cz/leaderboard/get.php");
        httpPost.setContent(HttpParametersUtils.convertHttpParameters(parameters));

        callback.connecting();

        Gdx.net.sendHttpRequest(httpPost, new Net.HttpResponseListener() {
            public String status;

            public void handleHttpResponse(Net.HttpResponse httpResponse) {

                if (isTerminated())
                    return;

                status = httpResponse.getResultAsString();

                //parse message
                Json json = new Json();
                LeaderBoardResult result = json.fromJson(LeaderBoardResult.class, status);

                if (result != null) {

                    if (result.getError().equals("")) {
                        callback.leaderboardResult(result);
                    } else if (result.getError().equals(RESPONSE_DB_CONNECTION_ERROR)) {
                        callback.connectionToDatabaseFailed();
                    } else if (result.getError().equals(RESPONSE_UNREGISTERED_USER)) {
                        callback.unregisteredUser();
                    } else {
                        callback.failure();
                    }


                } else {
                    callback.failure();
                }

            }

            public void failed(Throwable t) {
                if (isTerminated())
                    return;

                callback.connectionFailed();
            }

            @Override
            public void cancelled() {
                if (isTerminated())
                    return;

                callback.connectionCanceled();
            }
        });
    }


    public void register(String id, String nick, int level, String country, final LeaderBoardCallback callback) {
        terminated = false;

        Map parameters = new HashMap();
        parameters.put("id", id);
        parameters.put("nick", nick);
        parameters.put("level", String.valueOf(level));
        parameters.put("country", country);

        try {
            String hashString = id + nick + String.valueOf(level) + country + GameData.codeX;
            parameters.put("checksum", md5(hashString));
        } catch (NoSuchAlgorithmException e) {
            callback.failedToInitializeMD5();
        }

        Net.HttpRequest httpPost = new Net.HttpRequest(Net.HttpMethods.POST);
        httpPost.setUrl("http://sheepsgohome-com.svethostingu-tmp.cz/leaderboard/update.php");
        httpPost.setContent(HttpParametersUtils.convertHttpParameters(parameters));

        callback.connecting();

        Gdx.net.sendHttpRequest(httpPost, new Net.HttpResponseListener() {
            public String status;

            public void handleHttpResponse(Net.HttpResponse httpResponse) {
                if (isTerminated())
                    return;

                status = httpResponse.getResultAsString();

                //parse message
                Json json = new Json();
                String response = json.fromJson(String.class, status);

                if (response.equals(RESPONSE_DB_CONNECTION_ERROR)) {
                    callback.connectionToDatabaseFailed();
                } else if (response.equals(RESPONSE_INVALID_DATA)) {
                    callback.invalidData();
                } else if (response.equals(RESPONSE_NICK_ALREADY_IN_USE)) {
                    callback.nickAlreadyInUse();
                } else if (response.equals(RESPONSE_SUCCESS)) {
                    callback.success();
                } else if (response.equals(RESPONSE_FAILURE)) {
                    callback.failure();
                } else {
                    callback.failure();
                }


            }

            public void failed(Throwable t) {
                if (isTerminated())
                    return;

                callback.connectionFailed();
            }

            @Override
            public void cancelled() {
                if (isTerminated())
                    return;

                callback.connectionCanceled();
            }
        });
    }

    public String md5(String input) throws NoSuchAlgorithmException {
        String result = input;
        if (input != null) {
            MessageDigest md = MessageDigest.getInstance("MD5"); //or "SHA-1"
            md.update(input.getBytes());
            BigInteger hash = new BigInteger(1, md.digest());
            result = hash.toString(16);
            while (result.length() < 32) { //40 for SHA-1
                result = "0" + result;
            }
        }
        return result;
    }

    public synchronized boolean isTerminated() {
        return terminated;
    }

    public synchronized void setTerminated(boolean terminated) {
        this.terminated = terminated;
    }
}

File: core/src/com/sheepsgohome/leaderboard/LeaderBoard.kt
package com.sheepsgohome.leaderboard

import com.badlogic.gdx.Gdx
import com.badlogic.gdx.Net
import com.badlogic.gdx.net.HttpParametersUtils
import com.badlogic.gdx.utils.Json
import com.sheepsgohome.shared.GameData
import java.math.BigInteger
import java.security.MessageDigest
import java.security.NoSuchAlgorithmException

class LeaderBoard {

    companion object {
        val RESPONSE_DB_CONNECTION_ERROR = "DB_CONNECTION_ERROR"
        val RESPONSE_INVALID_DATA = "INVALID_DATA"
        val RESPONSE_NICK_ALREADY_IN_USE = "NICK_ALREADY_IN_USE"
        val RESPONSE_SUCCESS = "SUCCESS"
        val RESPONSE_FAILURE = "FAILURE"
        val RESPONSE_UNREGISTERED_USER = "UNREGISTERED_USER"

        val instance by lazy { LeaderBoard() }
    }

    var isTerminated: Boolean = false
        @Synchronized get
        @Synchronized set

    fun fetchLeaderboard(id: String, callback: LeaderBoardCallback) {
        isTerminated = false

        val parameters = mapOf("id" to id)

        val httpPost = Net.HttpRequest(Net.HttpMethods.POST)
        httpPost.url = "http://sheepsgohome-com.svethostingu-tmp.cz/leaderboard/get.php"
        httpPost.content = HttpParametersUtils.convertHttpParameters(parameters)

        callback.connecting()

        Gdx.net.sendHttpRequest(httpPost, object : Net.HttpResponseListener {
            var status: String = ""

            override fun handleHttpResponse(httpResponse: Net.HttpResponse) {

                if (isTerminated)
                    return

                status = httpResponse.resultAsString

                //parse message
                val json = Json()
                val result = json.fromJson(LeaderBoardResult::class.java, status)

                if (result != null) {

//                    when(result.error) {
//                        "" -> callback.leaderboardResult(result)
//                        RESPONSE_DB_CONNECTION_ERROR -> callback.connectionToDatabaseFailed()
//                        RESPONSE_UNREGISTERED_USER -> callback.unregisteredUser()
//                        else -> callback.failure()
//                    }


                } else {
                    callback.failure()
                }

            }

            override fun failed(t: Throwable) {
                if (isTerminated)
                    return

                callback.connectionFailed()
            }

            override fun cancelled() {
                if (isTerminated)
                    return

                callback.connectionCanceled()
            }
        })
    }

    fun register(id: String, nick: String, level: Int, country: String, callback: LeaderBoardCallback) {
        isTerminated = false

        val parameters = mutableMapOf(
            "id" to id,
            "nick" to nick,
            "level" to level.toString(),
            "country" to country
        )

        try {
            val hashString = id + nick + level.toString() + country + GameData.codeX
            parameters.put("checksum", md5(hashString))
        } catch (e: NoSuchAlgorithmException) {
            callback.failedToInitializeMD5()
        }

        val httpPost = Net.HttpRequest(Net.HttpMethods.POST)
        httpPost.url = "http://sheepsgohome-com.svethostingu-tmp.cz/leaderboard/update.php"
        httpPost.content = HttpParametersUtils.convertHttpParameters(parameters)

        callback.connecting()

        Gdx.net.sendHttpRequest(httpPost, object : Net.HttpResponseListener {
            var status: String = ""

            override fun handleHttpResponse(httpResponse: Net.HttpResponse) {
                if (isTerminated)
                    return

                status = httpResponse.resultAsString

                //parse message
                val json = Json()
                val response = json.fromJson(String::class.java, status)

                when (response) {
                    RESPONSE_DB_CONNECTION_ERROR -> callback.connectionToDatabaseFailed()
                    RESPONSE_INVALID_DATA -> callback.invalidData()
                    RESPONSE_NICK_ALREADY_IN_USE -> callback.nickAlreadyInUse()
                    RESPONSE_SUCCESS -> callback.success()
                    RESPONSE_FAILURE -> callback.failure()
                    else -> callback.failure()
                }
            }

            override fun failed(t: Throwable) {
                if (isTerminated)
                    return

                callback.connectionFailed()
            }

            override fun cancelled() {
                if (isTerminated)
                    return

                callback.connectionCanceled()
            }
        })
    }

    @Throws(NoSuchAlgorithmException::class)
    private fun md5(input: String?): String {
        var result: String = input ?: ""

        if (input != null) {
            val md = MessageDigest.getInstance("MD5") //or "SHA-1"
            md.update(input.toByteArray())
            val hash = BigInteger(1, md.digest())
            result = hash.toString(16)
            while (result.length < 32) { //40 for SHA-1
                result = "0" + result
            }
        }

        return result
    }

}

